### PR TITLE
Replace deprecated Array(T..., dims...) with Array{T...}(dims...) in /base/

### PR DESCRIPTION
**Pull request by pkofod**
Not quite done, but almost. All tests (besides some string-stuff) pass locally on ubuntu.

### CODE REVIEW

File: base/array.jl [hunk 1]
```diff
@@ -159,16 +159,16 @@ function fill!{T<:Union{Integer,AbstractFloat}}(a::Array{T}, x)
...
 end
 
-fill(v, dims::Dims)       = fill!(Array(typeof(v), dims), v)
-fill(v, dims::Integer...) = fill!(Array(typeof(v), dims...), v)
+fill(v, dims::Dims)       = fill!(Array{typeof(v)}(dims), v)
+fill(v, dims::Integer...) = fill!(Array{typeof(v)}(dims...), v)
 
-cell(dims::Integer...)   = Array(Any, dims...)
-cell(dims::Tuple{Vararg{Integer}}) = Array(Any, convert(Tuple{Vararg{Int}}, dims))
+cell(dims::Integer...)   = Array{Any}( dims...)
```
Comment (yuyichao) [inline base/array.jl]: Here's an extra space before `dims`

**Comment by yuyichao** [comment]
What are the string failure? Shouldn't the two constructors be equivalent?

**Comment by pkofod** [comment]
Sorry, I really explain myself. There was a string error before changing anything and they were there after as well. They seem to have been rebased away.

File: base/docs/helpdb/Base.jl [hunk 2]
```diff
@@ -7376,7 +7376,7 @@ for the effects of compilation.
     Array(dims)
 
 `Array{T}(dims)` constructs an uninitialized dense array with element type `T`. `dims` may
-be a tuple or a series of integer arguments. The syntax `Array(T, dims)` is also available,
+be a tuple or a series of integer arguments. The syntax `Array{T}(dims)` is also available,
```
Comment (mbauman) [inline base/docs/helpdb/Base.jl]: You don't want to change this one.  It's interesting that it's documented as deprecated, but I guess we never actually made it show a deprecation warning.

The next step will be to actually deprecate these: https://github.com/JuliaLang/julia/blob/99bf23db45241c10b6f2fd652e33aae60cad6059/base/boot.jl#L315-L320.  Running the tests with them deprecated will be an extra sanity check that you got them all.

Comment (pkofod) [inline base/docs/helpdb/Base.jl]: Yes of course, that is a mistake.

I will add the deprecation.

Comment (pkofod) [inline base/docs/helpdb/Base.jl]: Okay, I tried, and I have to say, I'm not sure how to do the deprecation. I understand it should be something like
```julia
@deprecate old new
```
but if I try something like
```julia
@deprecate Array{T,N}(::Type{T}, d::NTuple{N,Int}) = Array{T}(d)
```
I get something like (plus a lot more backtracing)
```jula
deprecated.jl
require.jl
error during bootstrap:
LoadError(at "sysimg.jl" line 337: LoadError(at "deprecated.jl" line 1186: Base.MethodError(f=Base.#@deprecate(), args=(Expr(:call, Expr(:curly, :Array, :T, :N)::Any, Expr(:::, Expr(:curly, :Type, :T)::Any)::Any, Expr(:::, :d, Expr(:curly, :NTuple, :N, :Int)::Any)::Any)::Any = Expr(:block, Expr(:line, 1186, :deprecated.jl)::Any, Expr(:call, Expr(:curly, :Array, :T)::Any, :d)::Any)::Any,))))
rec_backtrace at /home/pkm/julia5/julia/src/stackwalk.c:83
record_backtrace at /home/pkm/julia5/julia/src/task.c:228
jl_throw at /home/pkm/julia5/julia/src/task.c:514
eval at /home/pkm/julia5/julia/src/interpreter.c:415
jl_parse_eval_all at /home/pkm/julia5/julia/src/ast.c:773
jl_load at /home/pkm/julia5/julia/src/toplevel.c:583
#218 at ./sysimg.jl:188
jl_call_method_internal at /home/pkm/julia5/julia/src/julia_internal.h:78
do_call at /home/pkm/julia5/julia/src/interpreter.c:58
eval at /home/pkm/julia5/julia/src/interpreter.c:174
jl_toplevel_eval_flex at /home/pkm/julia5/julia/src/toplevel.c:545
jl_toplevel_eval_flex at /home/pkm/julia5/julia/src/toplevel.c:433
jl_toplevel_eval_flex at /home/pkm/julia5/julia/src/toplevel.c:443
jl_parse_eval_all at /home/pkm/julia5/julia/src/ast.c:777
jl_load at /home/pkm/julia5/julia/src/toplevel.c:583
exec_program at /home/pkm/julia5/julia/ui/repl.c:483
true_main at /home/pkm/julia5/julia/ui/repl.c:550
main at /home/pkm/julia5/julia/ui/repl.c:671
unknown function (ip: 0x7f7298be0a3f)
unknown function (ip: 0x401788)
``` 
or if I delete all the methods you linked to
```jlcon
docs/core.jl
inference.jl
A method error occurred before the base module was defined. Aborting...
Array
(Any, 9)
while loading inference.jl, in expression starting on line 3467
rec_backtrace at /home/pkm/julia5/julia/src/stackwalk.c:83
jl_method_error_bare at /home/pkm/julia5/julia/src/gf.c:965
jl_method_error at /home/pkm/julia5/julia/src/gf.c:976
jl_apply_generic at /home/pkm/julia5/julia/src/gf.c:1581
Type at ./inference.jl:92
jl_call_method_internal at /home/pkm/julia5/julia/src/julia_internal.h:78
typeinf_edge at ./inference.jl:1498
jl_call_method_internal at /home/pkm/julia5/julia/src/julia_internal.h:78
typeinf at ./inference.jl:1528
jl_call_method_internal at /home/pkm/julia5/julia/src/julia_internal.h:78
macro expansion; at ./inference.jl:3468
jl_call_method_internal at /home/pkm/julia5/julia/src/julia_internal.h:78
jl_parse_eval_all at /home/pkm/julia5/julia/src/ast.c:777
jl_load at /home/pkm/julia5/julia/src/toplevel.c:583
include at ./boot.jl:222
jl_call_method_internal at /home/pkm/julia5/julia/src/julia_internal.h:78
do_call at /home/pkm/julia5/julia/src/interpreter.c:58
eval at /home/pkm/julia5/julia/src/interpreter.c:174
jl_toplevel_eval_flex at /home/pkm/julia5/julia/src/toplevel.c:545
jl_toplevel_eval_flex at /home/pkm/julia5/julia/src/toplevel.c:433
jl_toplevel_eval_flex at /home/pkm/julia5/julia/src/toplevel.c:443
jl_toplevel_eval_in_warn at /home/pkm/julia5/julia/src/builtins.c:547
eval at ./boot.jl:225
jl_call_method_internal at /home/pkm/julia5/julia/src/julia_internal.h:78
do_call at /home/pkm/julia5/julia/src/interpreter.c:58
eval at /home/pkm/julia5/julia/src/interpreter.c:174
jl_toplevel_eval_flex at /home/pkm/julia5/julia/src/toplevel.c:545
jl_parse_eval_all at /home/pkm/julia5/julia/src/ast.c:777
jl_load at /home/pkm/julia5/julia/src/toplevel.c:583
exec_program at /home/pkm/julia5/julia/ui/repl.c:483
true_main at /home/pkm/julia5/julia/ui/repl.c:550
main at /home/pkm/julia5/julia/ui/repl.c:671
unknown function (ip: 0x7f81f1155a3f)
unknown function (ip: 0x401788)
Allocations: 298087 (Pool: 298048; Big: 39); GC: 0
```
I know I'm doing something wrong, I just can't figure out what.

Comment (yuyichao) [inline base/docs/helpdb/Base.jl]: `@deprecate Array{T,N}(::Type{T}, d::NTuple{N,Int}) Array{T}(d)` no `=`.

Comment (yuyichao) [inline base/docs/helpdb/Base.jl]: (The first error message says you are calling `@deprecate` with the wrong number of arguments), the second error message says you haven't fix the inference.jl for the deprecated constructor yet.

Comment (pkofod) [inline base/docs/helpdb/Base.jl]: > @deprecate Array{T,N}(::Type{T}, d::NTuple{N,Int}) Array{T}(d) no =.

Obviously, thanks. And thanks, I'll have a look at inference.jl.

Comment (yuyichao) [inline base/docs/helpdb/Base.jl]: (Please quote macros.... Apparently there are many creative usernames on github.....)

Comment (pkofod) [inline base/docs/helpdb/Base.jl]: > (Please quote macros.... Apparently there are many creative usernames on github.....)

(Sorry, didn't notice the quote was stripped when quoting using the "r" shortcut )

**Comment by tkelman** [comment]
This is great. I think we could merge it as is, and have a followup PR to add thr deprecation and clean up any last stragglers?

**Comment by pkofod** [comment]
Fine by me. I have not been outside of `base/`, so `test/` needs to be cleaned up as well.

**Comment by pkofod** [comment]
Some further changes. I still don't fully understand how I should do the deprecation. I can't really find any information about the content and structure of Inference.jl, and consequently I'm unsure what I need to change.

I take it the commits will simply be squashed upon merge, so I don't need to do that myself, yes?

**Comment by nalimilan** [comment]
I'm not sure why you want to modify inference.jl, but have a look at deprecated.jl.

**Comment by JeffBezanson** [comment]
Are we really sure we want to deprecate `Array(T, dims)`? Is there a discussion of this somebody can point me to?

I guess we will want the array constructor to be `Array{T}(iterable[, size])`, which will make it the same as Set, Dict, etc.

I also certainly hope inference.jl doesn't need to be modified for this.

**Comment by yuyichao** [comment]
As for the issue you see in type inference. It seems to be caused by type comprehension. `T[i for i in a]` is lowered to a call to `Array(T, n)` atm.

**Comment by pkofod** [comment]
> Are we really sure we want to deprecate Array(T, dims)? Is there a discussion of this somebody can point me to?

Are you referring to the changes in this PR currently, or if `Array(T, dims)` should still be allowed? I kinda of just noticed the deprecation note in the documentation when I was doing another PR, and figured I would clean the code according to the doc text. On the question of actually deprecating `Array(T, dims)` I guess this could be simply left out of this PR. It seems to be that you introduced the "deprecated" comment yourself
https://github.com/JuliaLang/julia/commit/23a6995a560ae88c14bf8e299faec7a4688df3f9 so I guess there hasn't necessarily been any discussion whether to actually deprecate it or not since then. 

> I'm not sure why you want to modify inference.jl, but have a look at deprecated.jl.

and

> I also certainly hope inference.jl doesn't need to be modified for this.

Well, if I deprecate the functions mbauman mentioned in comments from a previous commit, I get an error in `inference.jl` when `make`ing, which is apparantly due to


> As for the issue you see in type inference. It seems to be caused by type comprehension. T[i for i in a] is lowered to a call to Array(T, n) atm.

Regarding CI, it looks like a timeout at AppVeyor `x86_64`.

File: doc/manual/metaprogramming.rst [hunk 3]
```diff
@@ -102,7 +102,7 @@ objects:
     julia> dump(ex2)
     Expr
       head: Symbol call
-      args: Array(Any,(3,))
+      args: Array{Any}(3,))
```
Comment (tkelman) [inline doc/manual/metaprogramming.rst]: this is missing one `(`

**Comment by tkelman** [comment]
Maybe we don't have to deprecate the old way, but I think this is a bit better style and more generalizable to other array types without them all having to implement the "eltype as first positional argument" pattern. Besides, I think that pattern is a bit flawed - ref https://github.com/JuliaLang/julia/issues/11557.

File: doc/manual/constructors.rst [hunk 4]
```diff
@@ -586,7 +586,7 @@ to another, you should probably define a ``convert`` method instead.
 On the other hand, if your constructor does not represent a lossless
 conversion, or doesn't represent "conversion" at all, it is better
 to leave it as a constructor rather than a ``convert`` method.  For
-example, the ``Array(Int)`` constructor creates a zero-dimensional
+example, the ``Array{Int}`` constructor creates a zero-dimensional
```
Comment (tkelman) [inline doc/manual/constructors.rst]: this one is specifically documenting the paren method

Comment (pkofod) [inline doc/manual/constructors.rst]: Thanks for pointing it out. I will go through everything and look for these mistakes. I'm wondering, though. Should it be `Array(Int)` or `Array{Int}()` ? I've changed it to the latter in the actual code.

**Comment by StefanKarpinski** [comment]
I'm strongly in favor of deprecating the old way. Having too many ways to write things is just confusing.

**Comment by tkelman** [comment]
Let's hold off on making this pr any bigger and just merge it when green then, unless anyone sees any other cases that are confusing or typos. The deprecation can be a followup.

**Comment by pkofod** [comment]
Git error on AppVeyor.

**Comment by pkofod** [comment]
Not exactly sure what is happening with AppVeyor. If someone feels like it, you can stop the current runs on both CIs, and I can try to fire them again. I rebased twice within a short amount of time (I have a feeling that many commits are going to cause merge conflicts here... :) )

**Comment by pkofod** [comment]
Thanks. I'm not sure if you're doing this already @tkelman, but is it possible to stop the already running and queued runs ? ( https://travis-ci.org/JuliaLang/julia/pull_requests , https://ci.appveyor.com/project/JuliaLang/julia/history )

**Comment by tkelman** [comment]
There's some fail-fast code that automatically cancels them quickly when there are newer builds pending for the same PR.

File: doc/manual/interfaces.rst [hunk 5]
```diff
@@ -164,9 +164,8 @@ Methods to implement
 :func:`similar(A) <similar>`                                                 ``similar(A, eltype(A), size(A))``           Return a mutable array with the same shape and element type
 :func:`similar(A, ::Type{S}) <similar>`                                      ``similar(A, S, size(A))``                   Return a mutable array with the same shape and the specified element type
 :func:`similar(A, dims::NTuple{Int}) <similar>`                              ``similar(A, eltype(A), dims)``              Return a mutable array with the same element type and the specified dimensions
-:func:`similar(A, ::Type{S}, dims::NTuple{Int}) <similar>`                   ``Array(S, dims)``                           Return a mutable array with the specified element type and dimensions
+:func:`similar(A, ::Type{S}, dims::NTuple{Int}) <similar>`                   ``Array{S}(dims)``                           Return a mutable array with the specified element type and dimensions
 ============================================================================ ============================================ =======================================================================================
-
```
Comment (tkelman) [inline doc/manual/interfaces.rst]: sphinx might not like deleting this line but we can deal with that later if it causes any issues

**Pull request merged**